### PR TITLE
[PPS] 2021 conditions for timing detectors geometry (resubmission after DD4hep)

### PR DESCRIPTION
**Pull request by forthommel**
#### PR description:

This PR is a follow-up of #30252 after the merging of #31383 porting the PPS geometry into the DD4hep world. It includes a new set of DDLs for 2021 scenario (preparing for the addition of a new timing station), before integration onto the geometry DB payload.
The DetGeomDesc geometry parser is therefore modified to host this additional station, better exploiting the copy numbers in a modified hierarchy. As a result, this latter requires the addition of a "legacy" isRun2 flag to allow the previous DD/DB geometry format.

As a side development, the track reconstruction plugin is adapted to dynamically create a track fitting algorithm for each new pot/station observed in the rechits stream (i.e. timing pots are not hardcoded anymore in plugin implementation).
Algorithms being stored as an associative container (single pot detid → algorithm) of this latter, a hash function is defined for the CTPPSDetId object used as an indexing variable.

@fabferro @jan-kaspar @ghugo83 

#### PR validation:

Same validation procedure as #30996. Observed no difference in run 2 CTPPSDiamond* objects with the definition of the new station.
Overlaps detected (using the [g4OverlapCheck](https://github.com/cms-sw/cmssw/blob/master/SimG4Core/PrintGeomInfo/test/python/g4OverlapCheck_cfg.py) tool) equivalent to the ones already present in run 2 geometry (the current second station implementation is a naive copy-paste of this latter), and will be corrected in a follow-up PR. This PR being intended for an update of the reconstruction geometry payload (not the simulation one) it should not be problematic. Experts are warmly welcomed to comment.

#### if this PR is a backport please specify the original PR and why you need to backport that PR: N/A

Before submitting your pull requests, make sure you followed this checklist:
- [x] verify that the PR is really intended for the chosen branch
- [x] verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- [x] verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

File: Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc [hunk 1]
```diff
@@ -113,6 +115,7 @@ CTPPSGeometryESModule::CTPPSGeometryESModule(const edm::ParameterSet& iConfig)
 void CTPPSGeometryESModule::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
   edm::ParameterSetDescription desc;
   desc.addUntracked<unsigned int>("verbosity", 1);
+  desc.addUntracked<bool>("isRun2", false)->setComment("Switch to legacy (2017-18) definition of diamond geometry");
```
Comment (makortel) [inline Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc:118]: Should this parameter be tracked (it affects physics, no?)?

**Comment by civanch** [comment]
please test

File: Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc (hunk 1 above)
Comment (forthommel) [reply Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc:118]: Good point. Indeed it can be tracked (however this only affects run 2 configurations, so I believed such a 'legacy' behaviour could be hidden to the new configuration developments).
But maybe another FW expert can comment further?

**Comment by cvuosalo** [comment]
The comparison tests should differences in the Diamond Local Tracks. Is this change understood?

**Comment by forthommel** [comment]
> The comparison tests should differences in the Diamond Local Tracks. Is this change understood?

Not yet, I am investigating.

File: RecoPPS/Local/plugins/CTPPSDiamondLocalTrackFitter.cc [hunk 2]
```diff
@@ -28,73 +28,52 @@
...
-  trk_algo_56_.produceTracks(tracks56);
+  for (auto& algo_vs_id : trk_algo_) {
+    auto& tracks = pOut->find_or_insert(algo_vs_id.first);
+    // build the tracks for all stations
+    algo_vs_id.second->produceTracks(tracks);
+    // clear all hits to prepare for the next event
+    algo_vs_id.second->clear();
+  }
 
   iEvent.put(std::move(pOut));
```
Comment (perrotta) [inline RecoPPS/Local/plugins/CTPPSDiamondLocalTrackFitter.cc:76]: Shouldn't you put in the event **before** clearing the tracks?

**Comment by perrotta** [comment]
ping @forthommel : any news about your checks?

**Comment by silviodonato** [comment]
do you have any news? @forthommel

**Comment by forthommel** [comment]
@perrotta @silviodonato 
Sorry for the long reaction time! I am working on it, an update should appear in the following hours/days.

**Comment by perrotta** [comment]
> @perrotta @silviodonato
> Sorry for the long reaction time! I am working on it, an update should appear in the following hours/days.

Any news @forthommel ?

**Comment by civanch** [comment]
@forthommel , Perrotta and Matti pointed out to the problems, which seems to be not very difficult to fix.

**Comment by perrotta** [comment]
-1
- Since this PR seems to be unattended, let assign it a "-1" just to keep it out of the reco radars for the moment. Reco review will resume when any update will show up here

**Comment by silviodonato** [comment]
@forthommel do you have any news on this PR?

**Comment by silviodonato** [comment]
@forthommel any news?

**Comment by perrotta** [comment]
please test

**Comment by forthommel** [comment]
Sorry, forgot to push one commit!

**Comment by perrotta** [comment]
please test

**Comment by forthommel** [comment]
Very good! Only changes seem related to algo timing&logging services.

File: Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc (hunk 1 above)
Comment (cvuosalo) [reply Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc:118]: Why not make it tracked?

**Comment by cvuosalo** [comment]
The new file `Geometry/VeryForwardGeometry/data/dd4hep/geometryRPFromDD_2021.xml` should be called ` Geometry/VeryForwardGeometry/data/dd4hep/v1/geometryRPFromDD_2021.xml`.

**Comment by silviodonato** [comment]
@forthommel please take note also of https://github.com/cms-sw/cmssw/pull/31484#discussion_r489419050

**Comment by civanch** [comment]
@forthommel , note, that @cvuosalo comment is very important - when we already have a default 2021 geometry, each new update of any xml file should be a new version of this file.

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by forthommel** [comment]
Sorry for the extra commit, now ready for testing!

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

let us see the crash again

**Comment by civanch** [comment]
@forthommel , have you looked into this validation problem?

**Comment by forthommel** [comment]
@civanch (re-)looking at it now. All this mess for a simple untracked -> tracked parameter...

**Comment by forthommel** [comment]
Issue appears to be resolved on my test bench. Ready for a `please test`!

File: Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py [hunk 3]
```diff
@@ -1,12 +1,21 @@
...
                                                 )
 
-ctppsGeometryESModule = cms.ESProducer("CTPPSGeometryESModule",
-    verbosity = cms.untracked.uint32(1),
+ctppsGeometryESModule = CTPPSGeometryESModule.clone(
     compactViewTag = cms.string('XMLIdealGeometryESSource_CTPPS')
 )
+
+from Configuration.Eras.Modifier_ctpps_2016_cff import ctpps_2016
+ctpps_2016.toModify(ctppsGeometryESModule, isRun2=cms.bool(True))
```
Comment (makortel) [inline Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py:15]: It would be better to not add explicit `cms.bool()` here (in https://github.com/cms-sw/cmssw/commit/9b5a389cd3c5d23ab0e70a555cff2841b0cbab03)
```suggestion
ctpps_2016.toModify(ctppsGeometryESModule, isRun2=True)
```

Also, all the three could be combined along
```py
(ctpps_2016 | ctpps_2017 | ctpps_2018).toModify(ctppsGeometryESModule, isRun2=True)
```

Comment (forthommel) [reply Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py]: Many thanks, @makortel ! Committed your proposal in cba5238.

**Comment by forthommel** [comment]
Can someone please trigger the tests? Many thanks!

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by forthommel** [comment]
Okay, it seems that the import of an `ESSource` from an automatically-generated `cfi` file (using the `fillDescriptions` static member of the module) is introducing more complication than a good ole' copy-paste of a python `ESSource` definition... So be it!

**Comment by Dr15Jones** [comment]
The failure is
```
----- Begin Fatal Exception 01-Dec-2020 17:03:31 CET-----------------------
An exception of category 'EventSetupConflict' occurred while
   [0] Calling beginJob
Exception Message:
two EventSetup Producers want to deliver type="DetGeomDesc" label=""
 from record IdealGeometryRecord. The two providers are 
1) type="CTPPSGeometryESModule" label=""
2) type="CTPPSGeometryESModule" label="ctppsGeometryESModule"
Please either
   remove one of these Producers
   or find a way of configuring one of them so it does not deliver this data
   or use an es_prefer statement in the configuration to choose one.
----- End Fatal Exception -------------------------------------------------
```

Which means the module `CTPPSGeometryESModule ` was loaded twice in the configuration using two different labels.

**Comment by forthommel** [comment]
@Dr15Jones Yes indeed! I was treating these EventSetup modules the same way as "standard" modules which are only loaded in the runtime when part of a sequence or path. Maybe a bit of refactoring is required to avoid all these copy-paste definition of this `CTPPSGeometryESModule` all over our `cfi` files.

File: Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py [hunk 4]
```diff
@@ -1,12 +1,18 @@
 import FWCore.ParameterSet.Config as cms
+from Geometry.VeryForwardGeometryBuilder.CTPPSGeometryESModule_cfi import CTPPSGeometryESModule
```
Comment (Dr15Jones) [inline Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py]: Here is the problem. You are telling the system to also make the name `CTPPSGeometryESModule` available in this python module. 

Instead, do either
```python
import Geometry.VeryForwardGeometryBuilder.CTPPSGeometryESModule_cfi
```
and then only the clone line do
```python
ctppsGeometryESModule = Geometry.VeryForwardGeometryBuilder.CTPPSGeometryESModule_cfi.CTPPSGeometryESModule.clone(
```

or alternatively 
```python
from Geometry.VeryForwardGeometryBuilder.CTPPSGeometryESModule_cfi import CTPPSGeometryESModule as _ CTPPSGeometryESModule
```
and then the clone line becomes
```python
ctppsGeometryESModule =_CTPPSGeometryESModule.clone(
```

Comment (forthommel) [reply Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py]: Oh, did not know about this "underscoring" trick, many thanks @Dr15Jones! Maybe we can proceed as it is in 02dbab4cd08b7ee2943d6d077d0aa25dd0d971b4 and leave this cleanup for a follow-up PR?

Comment (forthommel) [reply Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py]: I meant 497ac85204102fe68bd3fda34e7659058cca3715. Sorry, now ready for testing.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
@forthommel , there is still problems.

File: Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py [hunk 5]
```diff
@@ -1,12 +1,19 @@
...
-                                                appendToDataLabel = cms.string('XMLIdealGeometryESSource_CTPPS')
-                                                )
+    rootDDName = cms.string('cms:CMSE'),
+    label = cms.string('CTPPS'),
+    appendToDataLabel = cms.string('XMLIdealGeometryESSource_CTPPS')
+)
 
 ctppsGeometryESModule = cms.ESProducer("CTPPSGeometryESModule",
     verbosity = cms.untracked.uint32(1),
+    isRun2 = False,
```
Comment (makortel) [inline Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py:11]: The definition of the ESProducer configuration needs the explicit type
```suggestion
    isRun2 = cms.bool(False),
```

Comment (forthommel) [reply Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py]: Many many thanks, @makortel! I was really starting to lose my sisu on this. So it seems the explicit type is indeed required for some parts. Should it also be replaced in the eras modifiers for safety? Or does https://github.com/cms-sw/cmssw/pull/31484#discussion_r533423845 still stand?

Comment (makortel) [reply Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py]: > So it seems the explicit type is indeed required for some parts.

The explicit types are needed in the definition of a module or `PSet`.

> Should it also be replaced in the eras modifiers for safety?

No, in all customizations of an existing module where an existing parameter is modified  (see below) it is better to leave out the explicit type (framework ensures that the parameter really exists and thus protects for e.g. typos, also less typing andrea ding)
```py
# direct override
ctppsGeometryESModule.isRun2 = False

# clone
anotherName = ctppsGeometryESModule.clone(isRun2 = False)

# modifier
someModifier.toModify(ctppsGeometryESModule, isRun2 = False)
```

In other words,

> Or does [#31484 (comment)](https://github.com/cms-sw/cmssw/pull/31484#discussion_r533423845) still stand?

yes, my earlier comment still stands.

> I was really starting to lose my sisu on this

Keep it up!

**Comment by makortel** [comment]
@cmsbuild, please test

File: Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py (hunk 5 above)
Comment (forthommel) [reply Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py]: Thanks a lot for the very detailed explanation! Much appreciated!

**Comment by forthommel** [comment]
Seems that the only failure relates to the pixel's `avgfedDigiOccvsLumi` of WF 8.0, so might be unrelated to this PR?

**Comment by forthommel** [comment]
Also, the difference in PPS collection sizes observed HI/MC workflows must arise from the new diamond tracking plugin which does not create both the arms' (empty) `edm::DetSet<CTPPSDiamondLocalTrack>` placeholders. They are now created "on the fly" when a new station is "detected" from the rechits stream. See [the corresponding code change](https://github.com/cms-sw/cmssw/pull/31484/files#diff-ff6dc5d9d54ec18f36f729089cd43c2e6a09d5ea7f8fbe3d51929d3d4556de02R62-R65).

**Comment by cvuosalo** [comment]
+1

**Comment by perrotta** [comment]
> Also, the difference in PPS collection sizes observed in HI/MC workflows must arise from the new diamond tracking plugin which does not create both the arms' (empty) `edm::DetSet<CTPPSDiamondLocalTrack>` placeholders. They are now created "on the fly" when a new station is "detected" from the rechits stream. See [the corresponding code change](https://github.com/cms-sw/cmssw/pull/31484/files#diff-ff6dc5d9d54ec18f36f729089cd43c2e6a09d5ea7f8fbe3d51929d3d4556de02R62-R65).

The differences observed in the test output show that the size of DiamondLocaTracks is 0 for MC, HI and 2016 real data. Can you confirm that this is what expected? IOW, could you please remind me exactly when those Diamond tracks were present? (And why they are not in the simulation, out of curiosity?)

**Comment by jan-kaspar** [comment]
The PPS diamond detectors were only installed in 2017.

AFAIK, at the moment, PPS does not have (full) simu for any Run2 wf.

Therefore, @perrotta 's observation seems expected to me. Obviously, @forthommel please correct me if I am wrong.

**Comment by forthommel** [comment]
Many thanks for the fast(er) reaction, @jan-kaspar! Indeed, 2 PPS diamond stations (1 per LHC "arm") were present from 2017 on. As mentioned above, the previous version of this plugin was ensuring at least these two stations had their `DetId` mapped in the `edm::DetSetVector<CTPPSDiamondLocalTrack>` collection, whereas a (tiny...) disk space is saved by only adding these two (and the additional stations to be installed by run 3) "on the fly" when they are discovered from the `CTPPSDiamondRecHit`s stream.
Therefore it is absolutely expected to have now completely empty collections for pre- and 2016 datasets, and of course for all simulated WFs (as no diamond/PPS in general simulation is run in the standard flows...)

**Comment by forthommel** [comment]
...which makes me wonder why the "simulation" signature is required for this PR...!

**Comment by perrotta** [comment]
> ...which makes me wonder why the "simulation" signature is required for this PR...!

 `DataFormats/CTPPSDetId` is under responsability of "simulation" (as well as all other  `DataFormats/*DetId`)

**Comment by forthommel** [comment]
Many thanks, @perrotta!
Now waiting for your, @slava77's, @jpata's, @ggovi's, @mdhildreth's and/or @civanch's signatures.

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
+1
- Only change in Reco is for the CTPPSDiamondLocalTrackFitter plugin, which now dynamically creates a track fitting algorithm for each new pot/station observed in the rechits stream (instead of hardcoding every time the two "45" and "56" ones, as it was done before): this will also allow dynamically taking into account of possible new ones
- Being they created dynamically, the size of those DiamondLocalTracks is now 0 when they are not present in the data or in the MC setup, as expected
- Jenkins tests do not show other differences

**Comment by forthommel** [comment]
Many thanks, @perrotta, @civanch, @cvuosalo!
@ggovi?

**Comment by forthommel** [comment]
@ggovi, let me know if I you need any input from me to ease your review!

**Comment by forthommel** [comment]
@ggovi ?

**Comment by forthommel** [comment]
@ggovi, please let us know if we can somehow help your review! This PR is currently a bottleneck and becoming more and more urgent in the scope of our future geometry/algorithmic developments!

**Comment by forthommel** [comment]
@perrotta, is @ggovi still in charge of the db signoff, by the way? I might be oppressing the wrong person :smile: 
If that is the case, apologies @ggovi! :)

**Comment by forthommel** [comment]
...and another round of signoffs!

**Comment by perrotta** [comment]
Why did you need to update @forthommel? Rebase needed?

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by forthommel** [comment]
> Why did you need to update @forthommel? Rebase needed?

There was an "unmergeable conflict" in `Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc` after the merging of https://github.com/cms-sw/cmssw/pull/32207...

**Comment by forthommel** [comment]
@perrotta although it was properly acknowledged by @cmsbuild, the IB seems to be lagging behind, isn't it?

File: CondTools/Geometry/plugins/PPSGeometryBuilder.cc [hunk 6]
```diff
@@ -37,6 +37,7 @@ class PPSGeometryBuilder : public edm::one::EDAnalyzer<> {
 
   bool fromDD4hep_;
   std::string compactViewTag_;
+  bool isRun2_;
```
Comment (ianna) [inline CondTools/Geometry/plugins/PPSGeometryBuilder.cc:40]: @forthommel - out of curiosity why cannot you rely on an iov to determine that?

File: Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py [hunk 7]
```diff
@@ -1,12 +1,19 @@
 import FWCore.ParameterSet.Config as cms
 
 XMLIdealGeometryESSource_CTPPS = cms.ESProducer("XMLIdealGeometryESProducer",
-                                                rootDDName = cms.string('cms:CMSE'),
-                                                label = cms.string('CTPPS'),
-                                                appendToDataLabel = cms.string('XMLIdealGeometryESSource_CTPPS')
-                                                )
+    rootDDName = cms.string('cms:CMSE'),
+    label = cms.string('CTPPS'),
```
Comment (ianna) [inline Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py:5]: @forthommel - it looks like we did not decouple this as we've discussed. I'm not sure this will continue working with DD4hep in production since we lock the main geometry after it has been parsed.

**Review by ianna — COMMENTED** [review]
@forthommel - I think, an agreement was that a schema for a `DetGeomDesc` would be developed and that it would become a RECO geometry in DB for PPS. I can still see DB reader configurations that firstly, do not rely on IOVs and secondly, still use XML parsing in RECO. Am I missing something? Thanks!

**Comment by perrotta** [comment]
@forthommel I'm curious: why this does not even build now? I supposed it had been just rebased with your last update...
In case you also applied other changes while rebasing, could you please list them here?

File: CondTools/Geometry/plugins/PPSGeometryBuilder.cc (hunk 6 above)
Comment (forthommel) [reply CondTools/Geometry/plugins/PPSGeometryBuilder.cc:40]: @ianna, thanks for your post-review! Indeed an IOV could have worked too, but as there is only one algorithmic change (i.e. a few hardcoded values becoming "dynamic") a simple flag is probably more efficient. But this discussion is indeed quite dated (3-4 months), and I cannot fully remember the various arguments (if any).

File: Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py (hunk 7 above)
Comment (forthommel) [reply Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py:5]: @ianna sorry for my ignorance. Is there any chance you still have a reference to the discussion thread so that I can see what could enter the scope of this PR (and what can possibly be introduced in a follow-up PR)?

**Comment by forthommel** [comment]
> @forthommel I'm curious: why this does not even build now? I supposed it had been just rebased with your last update...
> In case you also applied other changes while rebasing, could you please list them here?

I am also confused. I used the GH conflicts resolver interface which only listed one (strangely) unmerged path due to the introduction of the `bool isRun2_` member in the `CTPPSGeometryESModule` private attributes... Let me investigate!

File: CondTools/Geometry/plugins/PPSGeometryBuilder.cc (hunk 6 above)
Comment (jan-kaspar) [reply CondTools/Geometry/plugins/PPSGeometryBuilder.cc:40]: This was a long discussion. Eventually we decided to introduce this flag instead of re-uploading the Run2 geometry payloads.

File: Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py (hunk 7 above)
Comment (jan-kaspar) [reply Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py:5]: I am not sure what you mean by "decouple". I am not an expert for the DD4Hep option, but the result of the migration made by @ghugo83 seems to work. Do you expect any difference between the "private" tests and the application "in production"?

**Comment by jan-kaspar** [comment]
> @forthommel - I think, an agreement was that a schema for a `DetGeomDesc` would be developed and that it would become a RECO geometry in DB for PPS. I can still see DB reader configurations that firstly, do not rely on IOVs and secondly, still use XML parsing in RECO. Am I missing something? Thanks!

A short answer is that we are much slower that we like (however cannot do much differently given the manpower).

A longer answer: hopefully all our agreements are inside our TODO chart, cf. e.g. slide 3 in this GM presentation last week:
  https://indico.cern.ch/event/978940/contributions/4124285/attachments/2153246/3631224/kaspar_GM_reco_sw.pdf
The term "pre-processed" geometry refers to the RECO geometry (in binary format) in DB. Then, if you check the slides by @wpcarvalho
  https://indico.cern.ch/event/978940/contributions/4124295/attachments/2153652/3632031/PPS_DB_CMSWeek__2020-12-01.pdf
you'd see that this part is (almost) ready to be issued. It is just waiting for this PR and was waiting for #32207 merged recently.

Does this reply satisfactorily to your comments?

**Comment by forthommel** [comment]
Many many thanks for your help in this "fight", Jan! :)I'm really confused about this last compilation error, starting to suspect an IB underlying problem...Cheers,Laurent<div>
</div><div>
</div><div>-------- Original message --------</div><div>From: jan-kaspar <notifications@github.com> </div><div>Date: 09/12/2020  09:20  (GMT+01:00) </div><div>To: cms-sw/cmssw <cmssw@noreply.github.com> </div><div>Cc: Laurent Forthomme <laurent.forthomme@cern.ch>, Mention <mention@noreply.github.com> </div><div>Subject: Re: [cms-sw/cmssw] [PPS] 2021 conditions for timing detectors geometry (resubmission after DD4hep) (#31484) </div><div>
</div>>

**Comment by perrotta** [comment]
@forthommel the log of build error is quite exhaustive in pinpointing the origin of the error in https://github.com/cms-sw/cmssw/blob/3762cb0f063cf2fa4ca2cc1f7a576f93a6f1ecf3/CalibPPS/ESProducers/plugins/CTPPSCompositeESSource.cc#L262, which is a recent addition just merged with #32207
All newly inserted files PPS related must be checked and adapted to the modified interfaces in this PR

File: Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py (hunk 7 above)
Comment (ianna) [reply Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py:5]: @jan-kaspar - Detector description **by design is a source** to build desired geometries, it **must not** be a geometry itself. Neither `DD` nor `DD4hep` should allow usage of `XMLIdealGeometryESSource` as a RECO geometry. Besides, there is no guarantee it would work in a multi-threaded mode. It is `Xerces` implementation dependant, but neither `DocumentBuilderFactory` nor `DocumentBuilder` are guaranteed to be thread safe. It may work in a "private" test because you run in a single thread. It will most certainly fail "in production".

Comment (jan-kaspar) [reply Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py:5]: Thanks for the details!

Our plan is that in the next few days @wpcarvalho pushes a PR making PPS reco geometry being loaded from DB directly. The use of ``XMLIdealGeometryESSource`` (or so) will be then isolated to the preparation of the DB payloads.

**Comment by forthommel** [comment]
Thanks a lot @perrotta !
Now waiting for the next surprise :)

**Comment by forthommel** [comment]
In view to the ever-increasing number follow-up PRs depending on this one's merge, can it be justified that we get the "urgent" flag raised?

**Comment by perrotta** [comment]
please test

**Comment by silviodonato** [comment]
urgent

**Comment by forthommel** [comment]
`TestCTPPSDirectProtonSimulation` unit test is now fixed...

**Comment by silviodonato** [comment]
please test

File: Validation/CTPPS/python/simu_config/year_2016_cff.py [hunk 8]
```diff
@@ -12,11 +12,12 @@
 # geometry
 from Geometry.VeryForwardGeometry.geometryRPFromDD_2017_cfi import * # using 2017 here is OK
 ctppsCompositeESSource.compactViewTag = ctppsGeometryESModule.compactViewTag
+ctppsCompositeESSource.isRun2 = True
```
Comment (perrotta) [inline Validation/CTPPS/python/simu_config/year_2016_cff.py:15]: probably safer (here and below)
```suggestion
ctppsCompositeESSource.isRun2 = ctppsGeometryESModule.isRun2
```

Comment (forthommel) [reply Validation/CTPPS/python/simu_config/year_2016_cff.py:15]: @perrotta to pick up the flag configured by era?

**Comment by forthommel** [comment]
Maybe these tests can be aborted as a fix suggested in https://github.com/cms-sw/cmssw/pull/31484#discussion_r539292844 is ready to be pushed...

File: Validation/CTPPS/python/simu_config/year_2016_cff.py (hunk 8 above)
Comment (perrotta) [reply Validation/CTPPS/python/simu_config/year_2016_cff.py:15]: To pick up the very same `isRun2` flag used in `ctppsCompositeESSource`, as it is done right above for `compactViewTag`, and avoid possible mistakes when one has to modify by hand the same thing in several places.
(Minor comment, admittently)

File: Validation/CTPPS/python/simu_config/year_2016_cff.py [hunk 9]
```diff
@@ -12,11 +12,12 @@
...
 del ctppsGeometryESModule # this functionality is replaced by the composite ES source
```
Comment (makortel) [inline Validation/CTPPS/python/simu_config/year_2016_cff.py:16]: The preferred way to achieve this would be
```py
from Geometry.VeryForwardGeometry.geometryRPFromDD_2017_cfi import ctppsGeometryESModule as _ctppsGeometryESModule
... = _ctppsGeometryESModule.compactViewTag
```
Variables starting with underscore do not propagate with `import *` or `process.load()`.

**Comment by forthommel** [comment]
@makortel Thanks! The fix introduced in https://github.com/cms-sw/cmssw/pull/31484/commits/c0145c7f21cc0801afc5fd503aa332afb68ff984 is doing the job. However we may strip the ESSource (re)definition from these `geometryRPFromDD_*_cfi`s in (yet another) follow-up PR.

File: Validation/CTPPS/python/simu_config/year_2016_cff.py [hunk 10]
```diff
@@ -9,14 +9,16 @@
...
-# geometry
-from Geometry.VeryForwardGeometry.geometryRPFromDD_2017_cfi import * # using 2017 here is OK
-ctppsCompositeESSource.compactViewTag = ctppsGeometryESModule.compactViewTag
-del ctppsGeometryESModule # this functionality is replaced by the composite ES source
+# geometry (using 2017 here is OK)
+from Geometry.VeryForwardGeometry.geometryRPFromDD_2017_cfi import totemGeomXMLFiles, ctppsDiamondGeomXMLFiles, ctppsUFSDGeomXMLFiles, ctppsPixelGeomXMLFiles
+from Geometry.VeryForwardGeometry.geometryRPFromDD_2017_cfi import XMLIdealGeometryESSource_CTPPS
+from Geometry.VeryForwardGeometry.geometryRPFromDD_2017_cfi import ctppsGeometryESModule as _geom
+ctppsCompositeESSource.compactViewTag = _geom.compactViewTag
+ctppsCompositeESSource.isRun2 = _geom.isRun2
```
Comment (makortel) [inline Validation/CTPPS/python/simu_config/year_2016_cff.py:17]: Is this configuration fragment used in production or only in private tests? (by quick look it looks like the latter)

I'm asking because these lines modify a python object "owned" by (defined in) another python fragment. It may lead to surprising "long-distance interaction", like in this pathological example
```py
process.load("Path.To.fileThatDefines_ctppsCompositeESSource_cff")
import Path.To.fileThatDefines_ctppsCompositeESSource_cff
# process.ctppsCompositeESSource.isRun2 is now set as in year_2016_cff
```
Therefore in production configurations a python object is allowed to be modified only in the file that defines it.

Written that, for private test stuff where you have full control of everything, this could be ok, but be aware that the setup is somewhat fragile.

Comment (forthommel) [reply Validation/CTPPS/python/simu_config/year_2016_cff.py:17]: @jan-kaspar can confirm, but these are only used to define validation objects used in unit tests (hence these last minute modifications after the merge of #32207).

Comment (jan-kaspar) [reply Validation/CTPPS/python/simu_config/year_2016_cff.py:17]: The only "central" use is the unit test, as @forthommel writes.

Then, we (will) use these files extensively in PPS for "private" forward-proton simulations (cf. https://twiki.cern.ch/twiki/bin/viewauth/CMS/TaggedProtonsDirectSimulation). But I guess that this falls in the category "we have full control of everything".

**Comment by cvuosalo** [comment]
+1

**Comment by forthommel** [comment]
Seems that the comparisons are compatible with https://github.com/cms-sw/cmssw/pull/31484#issuecomment-727850242 (a few more WF seems to have appeared in the meantime though?)

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
+1
- Just rebased, and applied a few changes not reco related, since last reco signature in https://github.com/cms-sw/cmssw/pull/31484#issuecomment-737937800
- In the whole PR the only change in reco is for the `CTPPSDiamondLocalTrackFitter` plugin, which now dynamically creates a track fitting algorithm for each new pot/station observed in the rechits stream (instead of hardcoding every time the two "45" and "56" ones, as it was done before): this will also allow dynamically taking into account of possible new ones
- Being they created dynamically, the size of those DiamondLocalTracks is now 0 when they are not present in the data or in the MC setup, as expected (with respect to the previous comparisons, there are a few more MC workflows, and therefore a larger number of workflows with differences are reported by the automatic comparison)
- Jenkins tests do not show other differences

**Comment by forthommel** [comment]
Many thanks, @cvuosalo, @jfernan2, @perrotta!
@christopheralanwest, @tlampen, @pohsun, @yuanchao, @ggovi, @mdhildreth, @civanch?

**Comment by forthommel** [comment]
@christopheralanwest, @tlampen, @pohsun, @yuanchao, @ggovi, @mdhildreth, @civanch, as usual, let me/us know if you need any help to ease your review.

**Comment by ggovi** [comment]
+1

**Comment by forthommel** [comment]
Many thanks, @ggovi !
@christopheralanwest, @tlampen, @pohsun, @yuanchao, @mdhildreth, @civanch, kindly reminder :)

**Comment by silviodonato** [comment]
merge
@cms-sw/simulation-l2 already signed before the last rebase
@cms-sw/alca-l2 had to review only a trivial change `CalibPPS/ESProducers/plugins/CTPPSCompositeESSource.cc`
(please let us know if you have any comments)

I've opened an issue for a small followup PR regarding @makortel 's comment
https://github.com/cms-sw/cmssw/issues/32448

Even if `Validation/CTPPS/python/simu_config/year_201*_cff.py ` are currently used for a private production, they should be safe (see https://github.com/cms-sw/cmssw/pull/31484#discussion_r539393930)

**Comment by forthommel** [comment]
Thanks a lot, @silviodonato ! Off for a few more PRs now!

**Comment by yuanchao** [comment]
+1
